mlir/lib/Dialect/Rock/Transforms/RockMultibuffer.cpp

```diff
@@ -0,0 +1,452 @@
+  std::optional<OpFoldResult> singleStep = loop.getSingleStep();
+
+  if (!inductionVar || !lowerBound || !singleStep) {
+    llvm_unreachable("Loop variables should already have been tested.");
+  }
+  Value lbVal = getValueOrCreateConstantIndexOp(rewriter, loc, *lowerBound);
+  Value stepVal = getValueOrCreateConstantIndexOp(rewriter, loc, *singleStep);
+  Value ivVal = *inductionVar;
+  AffineExpr iv, lb, step;
+  bindDims(rewriter.getContext(), iv, lb, step);
+  Value bufferIndex = affine::makeComposedAffineApply(
+      rewriter, loc, ((iv - lb).floorDiv(step)) % multiBufferingFactor,
+      {ivVal, lbVal, stepVal});
+  LLVM_DEBUG(DBGS() << "--multi-buffered indexing: " << bufferIndex << "\n");
+  return bufferIndex;
+}
+
+/// Return true if the op fully overwrite the given `buffer` value.
+bool overrideBuffer(Operation *op, Value buffer) {
+  auto copyOp = dyn_cast<rock::RockWriterOpInterface>(op);
+  if (!copyOp)
+    return false;
+  auto maybeAllocOp = findAlloc(copyOp.getDest());
+  if (failed(maybeAllocOp))
+    return false;
+  return (*maybeAllocOp).getResult() == buffer;
+}
+
+/// Utility function to update the subtree rooted in the rock.gpuAlloc()
+/// As we go down the tree, we can meet
+///    a) Transformations (rock.transform)
+///    b) Operations that accept transformed views (e.g.,
+///    threadwise_read_into/threadwise_write_all) c) Operations that accept raw
+///    views (e.g., blockwise_gemm) d) memref.view to reinterpret the raw
+///    allocation and pass the buffer to other operations
+static void replaceUsesAndPropagateType(RewriterBase &rewriter, Location loc,
+                                        Operation *oldOp, Value val,
+                                        LoopLikeOpInterface loop,
+                                        int64_t loopLength,
+                                        int64_t multiBufferingFactor) {
+  SmallVector<Operation *> opsToDelete;
+
+  for (OpOperand &use : oldOp->getUses()) {
+    OpBuilder::InsertionGuard g(rewriter);
+    Operation *owner = use.getOwner();
+    rewriter.setInsertionPoint(owner);
+    // memref.view case: we expand the view type using the mbFactor. We call
+    // this expansion a reinterpret_multibuffer operation, that we can use
+    // to adjust the multibuffer factor at a later stage
+    if (auto view = dyn_cast<memref::ViewOp>(owner)) {
+      auto newView = rewriter.create<rock::ReinterpretMultiBufferOp>(
+          loc, val, view.getType(), multiBufferingFactor);
+      replaceUsesAndPropagateType(rewriter, loc, view, newView, loop,
+                                  loopLength, multiBufferingFactor);
+    } else if (auto transform = dyn_cast<TransformOp>(owner)) {
+      // rock.transform case      Value toTransform = val;
+      Value toTransform = val;
+      if (isTopOfTransformStack(val) && val.getType().isa<ShapedType>()) {
+        // if this is the top of the transform stack (i.e., the input to the
```

Comment:
I'm going to note that I've usually called this the bottom of the stack (see, for example, BottomUpTMBuilder taking the underlying shape as a starting point)

Reply:
... why resolved?